### ISSUE TITLE
RT-DETR onnx

### ISSUE THREAD

**Issue by lhj5426** [ultralytics/ultralytics#20327]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hello, I followed your suggestion today and trained a new model using RT-DETR, as mentioned in . 
https://github.com/ultralytics/ultralytics/issues/19523#issuecomment-2707481170
The results are indeed better than the previous YOLO model. However, I encountered a problem: the .PT file generated by the trained RT-DETR model cannot be converted to ONNX.

I tried using the previous YOLO .PY script for converting to ONNX, but it doesn't work for RT-DETR. Then I asked AI to write conversion code, but I couldn't get any working solutions from ChatGPT-4, Claude 3.7, or Deepseek.

### Additional

Could you please help me understand how to correctly convert it to ONNX? Thank you for your assistance!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @lhj5426, thank you for reaching out and for your interest in Ultralytics 🚀! This is an automated response to help you get started, and an Ultralytics engineer will also assist you soon.

We recommend checking out the [Docs](https://docs.ultralytics.com/) for guidance, including [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many common questions are already addressed there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it. This should include the specific commands you used, relevant code snippets, environment details, and any error messages.

If your question is about custom training or export, please include as much detail as possible, such as dataset examples, training logs, and the exact export commands/scripts you tried.

Join the Ultralytics community to connect with others and get real-time help:
- [Discord](https://discord.com/invite/ultralytics) 🎧 for chat
- [Discourse](https://community.ultralytics.com/) for in-depth discussions
- [Subreddit](https://reddit.com/r/Ultralytics) for community threads and tips

## Upgrade

Please upgrade to the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Environments

YOLO models, including RT-DETR, may be run in any of the following verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM ([GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/))
- **Amazon** Deep Learning AMI ([AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/))
- **Docker Image** ([Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/)) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you for providing as much detail as possible, and an Ultralytics engineer will review your question soon!

**Comment by Y-T-G** [comment]
What's the error? Can you provide the code you used?

**Comment by lhj5426** [comment]
> What's the error? Can you provide the code you used?

```
from ultralytics import RTDETR, YOLO
import os
import torch
import argparse

def export_rtdetr_to_onnx(model_path, output_path=None, half=False, simplify=True, opset=12):
    """
    使用Ultralytics API导出RT-DETR/YOLOv11模型为ONNX格式
    
    参数:
        model_path: 模型文件路径(.pt)
        output_path: 输出路径(可选，默认使用与模型同目录下的同名文件但后缀为.onnx)
        half: 是否使用FP16半精度
        simplify: 是否简化模型
        opset: ONNX操作集版本
    
    返回:
        导出的ONNX模型路径
    """
    print(f"正在加载模型: {model_path}")
    
    # 尝试使用RTDETR类加载
    try:
        model = RTDETR(model_path)
        print("使用RTDETR类加载模型成功")
    except Exception as e:
        print(f"使用RTDETR类加载失败: {e}")
        print("尝试使用YOLO类加载...")
        try:
            model = YOLO(model_path)
            print("使用YOLO类加载模型成功")
        except Exception as e:
            print(f"使用YOLO类加载也失败: {e}")
            return None
    
    # 定义输出路径
    if output_path is None:
        output_path = os.path.splitext(model_path)[0] + ".onnx"
    
    print(f"开始导出ONNX模型到: {output_path}")
    print(f"参数: half={half}, simplify={simplify}, opset={opset}")
    
    # 导出模型为ONNX格式
    try:
        success = model.export(format="onnx", 
                half=half,         # 是否使用FP16半精度
                simplify=simplify, # 是否简化模型
                dynamic=True,      # 支持动态输入尺寸
                opset=opset)       # ONNX操作集版本
        
        if success:
            print(f"ONNX模型导出成功: {output_path}")
            return output_path
        else:
            print("导出命令执行完毕，但返回值表示失败")
            return None
        
    except Exception as e:
        print(f"导出时发生错误: {e}")
        return None

def diagnostic_info():
    """输出诊断信息"""
    print("-" * 50)
    print("系统诊断信息:")
    
    try:
        import torch
        print(f"PyTorch版本: {torch.__version__}")
        print(f"CUDA可用: {torch.cuda.is_available()}")
        if torch.cuda.is_available():
            print(f"CUDA版本: {torch.version.cuda}")
            print(f"当前GPU: {torch.cuda.get_device_name(0)}")
    except ImportError:
        print("PyTorch未安装或导入失败")
    
    try:
        import onnx
        print(f"ONNX版本: {onnx.__version__}")
    except ImportError:
        print("ONNX未安装或导入失败")
    
    try:
        import ultralytics
        print(f"Ultralytics版本: {ultralytics.__version__}")
    except ImportError:
        print("Ultralytics未安装或导入失败")
    
    print("-" * 50)

if __name__ == "__main__":
    # 显示诊断信息
    diagnostic_info()
    
    # 解析命令行参数
    parser = argparse.ArgumentParser(description='导出RT-DETR/YOLOv11模型为ONNX格式')
    parser.add_argument('--model', type=str, default=r"J:\G\Desktop\0A1\best.pt", help='模型路径')
    parser.add_argument('--output', type=str, default=None, help='输出路径')
    parser.add_argument('--half', action='store_true', help='使用FP16半精度')
    parser.add_argument('--no-simplify', action='store_true', help='不简化模型')
    parser.add_argument('--opset', type=int, default=12, help='ONNX操作集版本')
    args = parser.parse_args()
    
    # 导出模型
    output_path = export_rtdetr_to_onnx(
        model_path=args.model,
        output_path=args.output,
        half=args.half,
        simplify=not args.no_simplify,
        opset=args.opset
    )
    
    if output_path:
        print(f"\n模型导出成功！")
        print(f"如果导出的模型无法使用，请尝试以下参数组合:")
        print(f"1. 不使用半精度, 操作集11: python script.py --model {args.model} --opset 11")
        print(f"2. 使用半精度, 操作集12: python script.py --model {args.model} --half --opset 12")
        print(f"3. 不简化模型: python script.py --model {args.model} --no-simplify --opset 12")
    else:
        print("\n导出失败，请尝试以下解决方案:")
        print("1. 更新Ultralytics库: pip install -U ultralytics")
        print("2. 安装或更新ONNX: pip install -U onnx onnxruntime")
        print("3. 如果您需要在特定框架使用ONNX模型，请考虑使用其他导出格式:")
        print("   - TorchScript: 通常更稳定，使用 --format=torchscript")
        print("   - TensorRT: 适用于NVIDIA GPU推理，使用 --format=engine")
```
This code was modified using 3 different AIs, but the conversion still failed in the end.

Do you need the model I trained for testing? I can provide the trained model.
https://alumnialbany-my.sharepoint.com/:u:/g/personal/planetrav_alumni_albany_edu/EaPVbsRDJjJLvKPJ50kxlyoBTC8XWJvs4HudoQoZk2vA5A?e=cvHBXp

**Comment by Y-T-G** [comment]
Why not just use `model.export()` with Ultralytics?

https://docs.ultralytics.com/modes/export/

**Comment by glenn-jocher** [comment]
Yes, our built-in export methods should work for RT-DETR models. The script you shared is unnecessarily complex - we have a simpler solution integrated directly in the API:

```python
from ultralytics import RTDETR

# Load your trained RT-DETR model
model = RTDETR("path/to/your/best.pt")

# Export to ONNX format with appropriate settings
model.export(format="onnx", dynamic=True, simplify=True)
```

This should generate an ONNX file in the same directory as your model. If you encounter any specific errors with this approach, please share the exact error message so we can troubleshoot it further.

For additional export options like adjusting the ONNX opset version or other parameters, you can check our [ONNX export documentation](https://docs.ultralytics.com/integrations/onnx/).
